src/Features/Core/Portable/EmbeddedLanguages/AbstractEmbeddedLanguageFeatureService.cs

```diff
@@ -7,101 +7,119 @@
 
-            foreach (var featureProvider in orderedFeatureProviders)
-            {
-                foreach (var identifier in featureProvider.Metadata.Identifiers)
-                    map.MultiAdd(identifier, featureProvider);
-            }
+        Info = info;
+        var languageIdentifiers = _identifierToServices.Keys.ToImmutableArray();
+        _detector = new EmbeddedLanguageDetector(info, languageIdentifiers, GetCommentDetector(languageIdentifiers));
 
-            foreach (var (_, services) in map)
-                services.RemoveDuplicates();
+        SyntaxTokenKinds.Add(syntaxKinds.CharacterLiteralToken);
+        SyntaxTokenKinds.Add(syntaxKinds.StringLiteralToken);
+        SyntaxTokenKinds.Add(syntaxKinds.InterpolatedStringTextToken);
 
-            this._identifierToServices = map.ToImmutableDictionary(
-                kvp => kvp.Key, kvp => kvp.Value.ToImmutableAndFree(), StringComparer.OrdinalIgnoreCase);
+        SyntaxTokenKinds.AddIfNotNull(syntaxKinds.SingleLineRawStringLiteralToken);
+        SyntaxTokenKinds.AddIfNotNull(syntaxKinds.MultiLineRawStringLiteralToken);
+        SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8StringLiteralToken);
+        SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8SingleLineRawStringLiteralToken);
+        SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8MultiLineRawStringLiteralToken);
+    }
 
-            Info = info;
-            _detector = new EmbeddedLanguageDetector(info, _identifierToServices.Keys.ToImmutableArray());
+    private static EmbeddedLanguageCommentDetector GetCommentDetector(ImmutableArray<string> languageIdentifiers)
+    {
+        // Well known language detectors we can cache.
 
-            SyntaxTokenKinds.Add(syntaxKinds.CharacterLiteralToken);
-            SyntaxTokenKinds.Add(syntaxKinds.StringLiteralToken);
-            SyntaxTokenKinds.Add(syntaxKinds.InterpolatedStringTextToken);
+        if (languageIdentifiers.SetEquals(JsonLanguageDetector.LanguageIdentifiers, StringComparer.OrdinalIgnoreCase))
+            return JsonLanguageDetector.CommentDetector;
 
-            SyntaxTokenKinds.AddIfNotNull(syntaxKinds.SingleLineRawStringLiteralToken);
-            SyntaxTokenKinds.AddIfNotNull(syntaxKinds.MultiLineRawStringLiteralToken);
-            SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8StringLiteralToken);
-            SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8SingleLineRawStringLiteralToken);
-            SyntaxTokenKinds.AddIfNotNull(syntaxKinds.Utf8MultiLineRawStringLiteralToken);
-        }
+        if (languageIdentifiers.SetEquals(RegexLanguageDetector.LanguageIdentifiers, StringComparer.OrdinalIgnoreCase))
+            return RegexLanguageDetector.CommentDetector;
+
+        if (languageIdentifiers.SetEquals(DateAndTimeLanguageDetector.LanguageIdentifiers, StringComparer.OrdinalIgnoreCase))
+            return DateAndTimeLanguageDetector.CommentDetector;
+
+        return new EmbeddedLanguageCommentDetector(languageIdentifiers);
```

Comment:
Maybe add a debug.assert here so we don't forget to update this when new embedded language support is added